File: src/ios/WindowState.m
#import "WindowState.h"

@implementation WindowState

typedef NS_ENUM(NSInteger, WindowStates) {
    Initialising, // 0
    Ready,        // 1
    Opening,      // 2
    Displayed,    // 3
    Hiding,       // 4
    Hidden,       // 5
    Unhiding,     // 6
    Closing,      // 7
    Exited        // 8
};

+ (NSString*)windowStateToString:(enum WindowStates) windowState
{
    NSString *result = nil;

    switch(windowState) {
        case Initialising:
            result = @"Initialising";
            break;
        case Ready:
            result = @"Ready";
            break;
        case Opening:
            result = @"Opening";
            break;
        case Displayed:
            result = @"Displayed";
            break;
        case Hiding:
            result = @"Hiding";
            break;
        case Hidden:
            result = @"Hidden";
            break;
        case Unhiding:
            result = @"Unhiding";
            break;
        case Closing:
            result = @"Closing";
            break;
        case Exited:
            result = @"Exited";
            break;
        default:
            [NSException raise:NSGenericException format:@"Unexpected WindowStates."];
    }

    return result;
}

WindowStates currentState;

// This makes the type a singleton.
+ (instancetype)sharedInstance {
    static WindowState* sharedInstance = nil;
    static dispatch_once_t onceToken;
    
    dispatch_once(&onceToken, ^{
        sharedInstance = [[WindowState alloc] init];
    });
    return sharedInstance;
}

- (id)init
{
    self = [super init];

    if (currentState != Initialising)
    {
        [self setState:Initialising];
    }

    return self;
}



- (void)setState:(WindowStates)newState
{
    WindowStates oldState = currentState;
    currentState = newState;
    NSLog(@"current window state changed from %@ to %@", [[self class] windowStateToString:oldState], [[self class] windowStateToString:newState]);
}

- (void)initialised
{
    [self setState:Ready];
}

- (void)close
{
    [self setState:Closing];
}

- (void)closed
{
    [self setState:Exited];
}

- (void)hide
{
    [self setState:Hiding];
}

- (void)hidden
{
    [self setState:Hidden];
}

- (void)unhide
{
    [self setState:Unhiding];
}

- (void)displayed
{
    [self setState:Displayed];
}

- (void)opening
{
    [self setState:Opening];
}

- (bool)canHide
{
    return currentState == Displayed || currentState == Unhiding || currentState == Hiding;
}

- (bool)canOpen
{
    return currentState == Ready;
}

- (bool)isHidden
{
   return currentState == Hidden;
}

- (bool)isUnhiding
{
   return currentState == Unhiding;
}
@end

File: src/android/WindowState.java
package org.apache.cordova.inappbrowser;

import org.apache.cordova.LOG;

public class WindowState {
    protected static final String LOG_TAG = "InAppBrowser";

    enum State {
        Initialising,
        Ready,
        Opening,
        Displayed,
        Hiding,
        HidingToBlank,
        Hidden,
        Unhiding,
        Closing,
        Exited
    }

    private static boolean reopenOnNextPageFinished = false;

    private static State currentState = State.Initialising;

    private static void setState(State newState) {
        LOG.d(LOG_TAG, String.format("current window state changed from %s to %s", currentState.name(), newState.name()));
        currentState = newState;
    }

    public static void ready() {
        setState(State.Ready);
    }

    public static void close() {
        setState(State.Closing);
    }

    public static void closed() {
        setState(State.Exited);
    }

    public static void hide() {
        setState(State.Hiding);
    }

    public static void hideToBlank() {
        setState(State.HidingToBlank);
    }

    public static void hidden() {
        setState(State.Hidden);
    }

    public static void unhide() {
        setState(State.Unhiding);
    }

    public static void displayed() {
        setState(State.Displayed);
    }

    public static void opening() {
        setState(State.Opening);
    }

    public static void reopenOnNextPageFinished() {
        reopenOnNextPageFinished = true;
    }

    public static void resetReopenOnNextPageFinished() {
        reopenOnNextPageFinished = false;
    }

// TODO: KPB - TBC, this is different to iOS
//    public static boolean canHide() {
//        return currentState == State.Displayed || currentState == State.Unhiding || currentState == State.Hiding;
//    }

    public static boolean canOpen() {
        // TODO: KPB - TBC, this is different to iOS
        return currentState == State.Ready || currentState == State.Hidden;
    }

    public static boolean isHidden() {
       return currentState == State.Hidden;
    }

    // TODO: KPB -this should be callled
    public static boolean isUnhiding() {
       return currentState == State.Unhiding;
    }

    public static boolean isHiding() {
        return currentState == State.Hiding || currentState == State.HidingToBlank;
    }

    public static boolean shouldHideBlank() {
        return currentState == State.HidingToBlank;
    }

    public static boolean shouldReopenOnNextPageFinished() {
        return reopenOnNextPageFinished;
    }
}